File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.java
package fr.free.nrw.commons.contributions;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.ImageView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentManager.OnBackStackChangedListener;
import androidx.work.ExistingWorkPolicy;
import com.google.android.material.bottomnavigation.BottomNavigationView.OnNavigationItemSelectedListener;
import fr.free.nrw.commons.databinding.FragmentContributionsListBinding;
import fr.free.nrw.commons.databinding.MainBinding;
import fr.free.nrw.commons.R;
import fr.free.nrw.commons.WelcomeActivity;
import fr.free.nrw.commons.auth.SessionManager;
import fr.free.nrw.commons.bookmarks.BookmarkFragment;
import fr.free.nrw.commons.explore.ExploreFragment;
import fr.free.nrw.commons.kvstore.JsonKvStore;
import fr.free.nrw.commons.location.LocationServiceManager;
import fr.free.nrw.commons.media.MediaClient;
import fr.free.nrw.commons.media.MediaDetailPagerFragment;
import fr.free.nrw.commons.navtab.MoreBottomSheetFragment;
import fr.free.nrw.commons.navtab.MoreBottomSheetLoggedOutFragment;
import fr.free.nrw.commons.navtab.NavTab;
import fr.free.nrw.commons.navtab.NavTabLayout;
import fr.free.nrw.commons.navtab.NavTabLoggedOut;
import fr.free.nrw.commons.nearby.Place;
import fr.free.nrw.commons.nearby.fragments.NearbyParentFragment;
import fr.free.nrw.commons.nearby.fragments.NearbyParentFragment.NearbyParentFragmentInstanceReadyCallback;
import fr.free.nrw.commons.notification.NotificationActivity;
import fr.free.nrw.commons.notification.NotificationController;
import fr.free.nrw.commons.profile.ProfileActivity;
import fr.free.nrw.commons.quiz.QuizChecker;
import fr.free.nrw.commons.settings.SettingsFragment;
import fr.free.nrw.commons.theme.BaseActivity;
import fr.free.nrw.commons.upload.UploadProgressActivity;
import fr.free.nrw.commons.upload.worker.WorkRequestHelper;
import fr.free.nrw.commons.utils.PermissionUtils;
import fr.free.nrw.commons.utils.ViewUtilWrapper;
import io.reactivex.Completable;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import java.util.Calendar;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import javax.inject.Inject;
import javax.inject.Named;
import kotlin.Unit;
import org.apache.commons.lang3.StringUtils;
import timber.log.Timber;

public class MainActivity extends BaseActivity
    implements OnBackStackChangedListener {

    @Inject
    SessionManager sessionManager;
    @Inject
    ContributionController controller;
    @Inject
    ContributionDao contributionDao;

    @Inject
    ContributionsListPresenter contributionsListPresenter;
    @Inject
    ContributionsRemoteDataSource dataSource;

    private ContributionsFragment contributionsFragment;
    private NearbyParentFragment nearbyParentFragment;
    private ExploreFragment exploreFragment;
    private BookmarkFragment bookmarkFragment;
    public ActiveFragment activeFragment;
    private MediaDetailPagerFragment mediaDetailPagerFragment;
    private OnNavigationItemSelectedListener navListener;

    @Inject
    public LocationServiceManager locationManager;
    @Inject
    NotificationController notificationController;
    @Inject
    QuizChecker quizChecker;
    @Inject
    @Named("default_preferences")
    public JsonKvStore applicationKvStore;
    @Inject
    ViewUtilWrapper viewUtilWrapper;

    public Menu menu;

    public MainBinding binding;

    NavTabLayout tabLayout;

    private FragmentContributionsListBinding refresh;

    private String userName;

    /**
     * Consumers should be simply using this method to use this activity.
     *
     * @param context A Context of the application package implementing this class.
     */
    public static void startYourself(Context context) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT | Intent.FLAG_ACTIVITY_SINGLE_TOP);
        context.startActivity(intent);
    }

    @Override
    public boolean onSupportNavigateUp() {
        if (activeFragment == ActiveFragment.CONTRIBUTIONS) {
            if (!contributionsFragment.backButtonClicked()) {
                return false;
            }
        } else {
            onBackPressed();
            showTabs();
        }
        return true;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = MainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.toolbarBinding.toolbar);
        tabLayout = binding.fragmentMainNavTabLayout;
        loadLocale();

        binding.toolbarBinding.toolbar.setNavigationOnClickListener(view -> {
            onSupportNavigateUp();
        });

        applicationKvStore.putBoolean("first_edit_depict", true);
        if (applicationKvStore.getBoolean("login_skipped") == true) {
            setTitle(getString(R.string.navigation_item_explore));
            setUpLoggedOutPager();
        } else {
            if (applicationKvStore.getBoolean("firstrun", true)) {
                applicationKvStore.putBoolean("hasAlreadyLaunchedBigMultiupload", false);
                applicationKvStore.putBoolean("hasAlreadyLaunchedCategoriesDialog", false);
            }
            if (savedInstanceState == null) {
                if (applicationKvStore.getBoolean("last_opened_nearby")) {
                    setTitle(getString(R.string.nearby_fragment));
                    showNearby();
                    loadFragment(NearbyParentFragment.newInstance(), false);
                } else {
                    setTitle(getString(R.string.contributions_fragment));
                    loadFragment(ContributionsFragment.newInstance(), false);
                }
            }
            refresh = FragmentContributionsListBinding.inflate(getLayoutInflater());
            if (getIntent().getExtras() != null) {
                userName = getIntent().getExtras().getString(ProfileActivity.KEY_USERNAME);
            }

            if (StringUtils.isEmpty(userName)) {
                userName = sessionManager.getUserName();
            }
            setUpPager();
            checkAndResumeStuckUploads();
        }
    }

    public void setSelectedItemId(int id) {
        binding.fragmentMainNavTabLayout.setSelectedItemId(id);
    }

    private void setUpPager() {
        binding.fragmentMainNavTabLayout.setOnNavigationItemSelectedListener(
            navListener = (item) -> {
                if (!item.getTitle().equals(getString(R.string.more))) {
                    setTitle(item.getTitle());
                }
                applicationKvStore.putBoolean("last_opened_nearby",
                    item.getTitle().equals(getString(R.string.nearby_fragment)));
                final Fragment fragment = NavTab.of(item.getOrder()).newInstance();
                return loadFragment(fragment, true);
            });
    }

    private void setUpLoggedOutPager() {
        loadFragment(ExploreFragment.newInstance(), false);
        binding.fragmentMainNavTabLayout.setOnNavigationItemSelectedListener(item -> {
            if (!item.getTitle().equals(getString(R.string.more))) {
                setTitle(item.getTitle());
            }
            Fragment fragment = NavTabLoggedOut.of(item.getOrder()).newInstance();
            return loadFragment(fragment, true);
        });
    }

    private boolean loadFragment(Fragment fragment, boolean showBottom) {
        if (fragment instanceof ContributionsFragment) {
            if (activeFragment == ActiveFragment.CONTRIBUTIONS) {
                contributionsFragment.scrollToTop();
                return true;
            }
            contributionsFragment = (ContributionsFragment) fragment;
            activeFragment = ActiveFragment.CONTRIBUTIONS;
        } else if (fragment instanceof NearbyParentFragment) {
            if (activeFragment == ActiveFragment.NEARBY) {
                return true;
            }
            nearbyParentFragment = (NearbyParentFragment) fragment;
            activeFragment = ActiveFragment.NEARBY;
        } else if (fragment instanceof ExploreFragment) {
            if (activeFragment == ActiveFragment.EXPLORE) {
                return true;
            }
            exploreFragment = (ExploreFragment) fragment;
            activeFragment = ActiveFragment.EXPLORE;
        } else if (fragment instanceof BookmarkFragment) {
            if (activeFragment == ActiveFragment.BOOKMARK) {
                return true;
            }
            bookmarkFragment = (BookmarkFragment) fragment;
            activeFragment = ActiveFragment.BOOKMARK;
        } else if (fragment == null && showBottom) {
            if (applicationKvStore.getBoolean("login_skipped") == true) {
                MoreBottomSheetLoggedOutFragment bottomSheet = new MoreBottomSheetLoggedOutFragment();
                bottomSheet.show(getSupportFragmentManager(), "MoreBottomSheetLoggedOut");
            } else {
                MoreBottomSheetFragment bottomSheet = new MoreBottomSheetFragment();
                bottomSheet.show(getSupportFragmentManager(), "MoreBottomSheet");
            }
        }

        if (fragment != null) {
            getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.fragmentContainer, fragment)
                .commit();
            return true;
        }
        return false;
    }

    public void hideTabs() {
        binding.fragmentMainNavTabLayout.setVisibility(View.GONE);
    }

    public void showTabs() {
        binding.fragmentMainNavTabLayout.setVisibility(View.VISIBLE);
    }

    public void setNumOfUploads(int uploadCount) {
        if (activeFragment == ActiveFragment.CONTRIBUTIONS) {
            setTitle(getResources().getString(R.string.contributions_fragment) + " " + (
                !(uploadCount == 0) ?
                    getResources()
                        .getQuantityString(R.plurals.contributions_subtitle,
                            uploadCount, uploadCount)
                    : getString(R.string.contributions_subtitle_zero)));
        }
    }

    @SuppressLint("CheckResult")
    private void checkAndResumeStuckUploads() {
        List<Contribution> stuckUploads = contributionDao.getContribution(
                Collections.singletonList(Contribution.STATE_IN_PROGRESS))
            .subscribeOn(Schedulers.io())
            .blockingGet();
        Timber.d("Resuming " + stuckUploads.size() + " uploads...");
        if (!stuckUploads.isEmpty()) {
            for (Contribution contribution : stuckUploads) {
                contribution.setState(Contribution.STATE_QUEUED);
                contribution.setDateUploadStarted(Calendar.getInstance().getTime());
                Completable.fromAction(() -> contributionDao.saveSynchronous(contribution))
                    .subscribeOn(Schedulers.io())
                    .subscribe();
            }
            WorkRequestHelper.Companion.makeOneTimeWorkRequest(
                this, ExistingWorkPolicy.APPEND_OR_REPLACE);
        }
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("viewPagerCurrentItem", binding.pager.getCurrentItem());
        outState.putString("activeFragment", activeFragment.name());
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        String activeFragmentName = savedInstanceState.getString("activeFragment");
        if (activeFragmentName != null) {
            restoreActiveFragment(activeFragmentName);
        }
    }

    private void restoreActiveFragment(@NonNull String fragmentName) {
        if (fragmentName.equals(ActiveFragment.CONTRIBUTIONS.name())) {
            setTitle(getString(R.string.contributions_fragment));
            loadFragment(ContributionsFragment.newInstance(), false);
        } else if (fragmentName.equals(ActiveFragment.NEARBY.name())) {
            setTitle(getString(R.string.nearby_fragment));
            loadFragment(NearbyParentFragment.newInstance(), false);
        } else if (fragmentName.equals(ActiveFragment.EXPLORE.name())) {
            setTitle(getString(R.string.navigation_item_explore));
            loadFragment(ExploreFragment.newInstance(), false);
        } else if (fragmentName.equals(ActiveFragment.BOOKMARK.name())) {
            setTitle(getString(R.string.bookmarks));
            loadFragment(BookmarkFragment.newInstance(), false);
        }
    }

    @Override
    public void onBackPressed() {
        if (contributionsFragment != null && activeFragment == ActiveFragment.CONTRIBUTIONS) {
            if (!contributionsFragment.backButtonClicked()) {
                super.onBackPressed();
            }
        } else if (nearbyParentFragment != null && activeFragment == ActiveFragment.NEARBY) {
            if (!nearbyParentFragment.backButtonClicked()) {
                getSupportFragmentManager().beginTransaction().remove(nearbyParentFragment)
                    .commit();
                setSelectedItemId(NavTab.CONTRIBUTIONS.code());
            }
        } else if (exploreFragment != null && activeFragment == ActiveFragment.EXPLORE) {
            if (!exploreFragment.onBackPressed()) {
                if (applicationKvStore.getBoolean("login_skipped")) {
                    super.onBackPressed();
                } else {
                    setSelectedItemId(NavTab.CONTRIBUTIONS.code());
                }
            }
        } else if (bookmarkFragment != null && activeFragment == ActiveFragment.BOOKMARK) {
            bookmarkFragment.onBackPressed();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public void onBackStackChanged() {
    }

    @SuppressLint("CheckResult")
    private void retryAllFailedUploads() {
        contributionDao.
            getContribution(Collections.singletonList(Contribution.STATE_FAILED))
            .subscribeOn(Schedulers.io())
            .subscribe(failedUploads -> {
                for (Contribution contribution : failedUploads) {
                    contributionsFragment.retryUpload(contribution);
                }
            });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.upload_tab:
                startActivity(new Intent(this, UploadProgressActivity.class));
                return true;
            case R.id.notifications:
                NotificationActivity.Companion.startYourself(this, "unread");
                return true;
            case R.id.menu_refresh:
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.contribution_activity_notification_menu, menu);

        MenuItem refreshItem = menu.findItem(R.id.menu_refresh);
        if (refreshItem != null) {
            View actionView = refreshItem.getActionView();
            if (actionView != null) {
                ImageView refreshIcon = actionView.findViewById(R.id.refresh_icon);
                if (refreshIcon != null) {
                    refreshIcon.setOnClickListener(v -> {
                        v.clearAnimation();
                        Animation rotateAnimation = AnimationUtils.loadAnimation(this, R.anim.rotate);
                        v.startAnimation(rotateAnimation);

                        // Initialize userName if it's null
                        if (userName == null) {
                            userName = sessionManager.getUserName();
                        }

                        if (Objects.equals(sessionManager.getUserName(), userName)) {
                            if (refresh != null && refresh.swipeRefreshLayout != null) {
                                refresh.swipeRefreshLayout.setRefreshing(true);
                                refresh.swipeRefreshLayout.setEnabled(true);
                                contributionsListPresenter.refreshList(refresh.swipeRefreshLayout);
                            }
                        } else {
                            if (refresh != null && refresh.swipeRefreshLayout != null) {
                                refresh.swipeRefreshLayout.setEnabled(false);
                            }
                        }
                    });
                }
            }
        }

        return true;
    }

    public void centerMapToPlace(Place place) {
        setSelectedItemId(NavTab.NEARBY.code());
        nearbyParentFragment.setNearbyParentFragmentInstanceReadyCallback(
            new NearbyParentFragmentInstanceReadyCallback() {
                @Override
                public void onReady() {
                    nearbyParentFragment.centerMapToPlace(place);
                }
            });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (applicationKvStore.getBoolean("firstrun", true) &&
            !applicationKvStore.getBoolean("login_skipped")) {
            defaultKvStore.putBoolean("inAppCameraFirstRun", true);
            WelcomeActivity.startYourself(this);
        }

        retryAllFailedUploads();
        //check for new contributions
        // Initialize userName if it's null
        if (userName == null) {
            userName = sessionManager.getUserName();
        }

        if (Objects.equals(sessionManager.getUserName(), userName)) {
            if (refresh != null && refresh.swipeRefreshLayout != null) {
                refresh.swipeRefreshLayout.setRefreshing(true);
                refresh.swipeRefreshLayout.setEnabled(true);
                contributionsListPresenter.refreshList(refresh.swipeRefreshLayout);
            }
        } else {
            if (refresh != null && refresh.swipeRefreshLayout != null) {
                refresh.swipeRefreshLayout.setEnabled(false);
            }
        }
    }

    @Override
    protected void onDestroy() {
        quizChecker.cleanup();
        locationManager.unregisterLocationManager();
        locationManager = null;
        super.onDestroy();
    }

    public void showNearby() {
        binding.fragmentMainNavTabLayout.setSelectedItemId(NavTab.NEARBY.code());
    }

    public enum ActiveFragment {
        CONTRIBUTIONS,
        NEARBY,
        EXPLORE,
        BOOKMARK,
        MORE
    }

    private void loadLocale() {
        final SharedPreferences preferences = getSharedPreferences("Settings",
            Activity.MODE_PRIVATE);
        final String language = preferences.getString("language", "");
        final SettingsFragment settingsFragment = new SettingsFragment();
        settingsFragment.setLocale(this, language);
    }

    public OnNavigationItemSelectedListener getNavListener() {
        return navListener;
    }
}

File: app/src/main/java/fr/free/nrw/commons/contributions/MainActivity.kt
package fr.free.nrw.commons.contributions

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.work.ExistingWorkPolicy
import com.google.android.material.bottomnavigation.BottomNavigationView
import fr.free.nrw.commons.R
import fr.free.nrw.commons.auth.SessionManager
import fr.free.nrw.commons.bookmarks.BookmarkFragment
import fr.free.nrw.commons.contributions.ContributionsFragment.Companion.newInstance
import fr.free.nrw.commons.databinding.MainBinding
import fr.free.nrw.commons.explore.ExploreFragment
import fr.free.nrw.commons.kvstore.JsonKvStore
import fr.free.nrw.commons.location.LocationServiceManager
import fr.free.nrw.commons.media.MediaDetailPagerFragment
import fr.free.nrw.commons.navtab.MoreBottomSheetFragment
import fr.free.nrw.commons.navtab.MoreBottomSheetLoggedOutFragment
import fr.free.nrw.commons.navtab.NavTab
import fr.free.nrw.commons.navtab.NavTabLayout
import fr.free.nrw.commons.navtab.NavTabLoggedOut
import fr.free.nrw.commons.nearby.Place
import fr.free.nrw.commons.nearby.fragments.NearbyParentFragment
import fr.free.nrw.commons.nearby.fragments.NearbyParentFragment.NearbyParentFragmentInstanceReadyCallback
import fr.free.nrw.commons.notification.NotificationActivity.Companion.startYourself
import fr.free.nrw.commons.notification.NotificationController
import fr.free.nrw.commons.quiz.QuizChecker
import fr.free.nrw.commons.settings.SettingsFragment
import fr.free.nrw.commons.startWelcome
import fr.free.nrw.commons.theme.BaseActivity
import fr.free.nrw.commons.utils.applyEdgeToEdgeAllInsets
import fr.free.nrw.commons.upload.UploadProgressActivity
import fr.free.nrw.commons.upload.worker.WorkRequestHelper.Companion.makeOneTimeWorkRequest
import fr.free.nrw.commons.utils.ViewUtilWrapper
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.Completable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import java.util.Calendar
import javax.inject.Inject
import javax.inject.Named

@AndroidEntryPoint
class MainActivity : BaseActivity(), FragmentManager.OnBackStackChangedListener {
    @JvmField
    @Inject
    var sessionManager: SessionManager? = null

    @JvmField
    @Inject
    var controller: ContributionController? = null

    @JvmField
    @Inject
    var contributionDao: ContributionDao? = null

    private var contributionsFragment: ContributionsFragment? = null
    private var nearbyParentFragment: NearbyParentFragment? = null
    private var exploreFragment: ExploreFragment? = null
    private var bookmarkFragment: BookmarkFragment? = null
    @JvmField
    var activeFragment: ActiveFragment? = null
    private val mediaDetailPagerFragment: MediaDetailPagerFragment? = null
    var navListener: BottomNavigationView.OnNavigationItemSelectedListener? = null
        private set

    @JvmField
    @Inject
    var locationManager: LocationServiceManager? = null

    @JvmField
    @Inject
    var notificationController: NotificationController? = null

    @JvmField
    @Inject
    var quizChecker: QuizChecker? = null

    @JvmField
    @Inject
    @Named("default_preferences")
    var applicationKvStore: JsonKvStore? = null

    @JvmField
    @Inject
    var viewUtilWrapper: ViewUtilWrapper? = null

    var menu: Menu? = null

    @JvmField
    var binding: MainBinding? = null

    var tabLayout: NavTabLayout? = null


    override fun onSupportNavigateUp(): Boolean {
        if (activeFragment == ActiveFragment.CONTRIBUTIONS) {
            if (!contributionsFragment!!.backButtonClicked()) {
                return false
            }
        } else {
            onBackPressed()
            showTabs()
        }
        return true
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = MainBinding.inflate(layoutInflater)
        applyEdgeToEdgeAllInsets(binding!!.root)
        setContentView(binding!!.root)
        setSupportActionBar(binding!!.toolbarBinding.toolbar)
        tabLayout = binding!!.fragmentMainNavTabLayout
        loadLocale()

        binding!!.toolbarBinding.toolbar.setNavigationOnClickListener { view: View? ->
            onSupportNavigateUp()
        }
        /*
"first_edit_depict" is a key for getting information about opening the depiction editor
screen for the first time after opening the app.

Getting true by the key means the depiction editor screen is opened for the first time
after opening the app.
Getting false by the key means the depiction editor screen is not opened for the first time
after opening the app.
 */
        applicationKvStore!!.putBoolean("first_edit_depict", true)
        if (applicationKvStore!!.getBoolean("login_skipped") == true) {
            title = getString(R.string.navigation_item_explore)
            setUpLoggedOutPager()
        } else {
            if (applicationKvStore!!.getBoolean("firstrun", true)) {
                applicationKvStore!!.putBoolean("hasAlreadyLaunchedBigMultiupload", false)
                applicationKvStore!!.putBoolean("hasAlreadyLaunchedCategoriesDialog", false)
            }
            if (savedInstanceState == null) {
                //starting a fresh fragment.
                // Open Last opened screen if it is Contributions or Nearby, otherwise Contributions
                if (applicationKvStore!!.getBoolean("last_opened_nearby")) {
                    title = getString(R.string.nearby_fragment)
                    showNearby()
                    loadFragment(NearbyParentFragment.newInstance(), false)
                } else {
                    title = getString(R.string.contributions_fragment)
                    loadFragment(newInstance(), false)
                }
            }
            setUpPager()

            checkAndResumeStuckUploads()
        }
    }

    fun setSelectedItemId(id: Int) {
        binding!!.fragmentMainNavTabLayout.selectedItemId = id
    }

    private fun setUpPager() {
        binding!!.fragmentMainNavTabLayout.setOnNavigationItemSelectedListener(
            BottomNavigationView.OnNavigationItemSelectedListener { item: MenuItem ->
                if (item.title != getString(R.string.more)) {
                    // do not change title for more fragment
                    title = item.title
                }
                // set last_opened_nearby true if item is nearby screen else set false
                applicationKvStore!!.putBoolean(
                    "last_opened_nearby",
                    item.title == getString(R.string.nearby_fragment)
                )
                val fragment = NavTab.of(item.order).newInstance()
                loadFragment(fragment, true)
            }.also { navListener = it })
    }

    private fun setUpLoggedOutPager() {
        loadFragment(ExploreFragment.newInstance(), false)
        binding!!.fragmentMainNavTabLayout.setOnNavigationItemSelectedListener { item: MenuItem ->
            if (item.title != getString(R.string.more)) {
                // do not change title for more fragment
                title = item.title
            }
            val fragment =
                NavTabLoggedOut.of(item.order).newInstance()
            loadFragment(fragment, true)
        }
    }

    private fun loadFragment(fragment: Fragment?, showBottom: Boolean): Boolean {
        //showBottom so that we do not show the bottom tray again when constructing
        //from the saved instance state.

        freeUpFragments();

        if (fragment is ContributionsFragment) {
            if (activeFragment == ActiveFragment.CONTRIBUTIONS) {
                // scroll to top if already on the Contributions tab
                contributionsFragment!!.scrollToTop()
                return true
            }
            contributionsFragment = fragment
            activeFragment = ActiveFragment.CONTRIBUTIONS
        } else if (fragment is NearbyParentFragment) {
            if (activeFragment == ActiveFragment.NEARBY) { // Do nothing if same tab
                return true
            }
            nearbyParentFragment = fragment
            activeFragment = ActiveFragment.NEARBY
        } else if (fragment is ExploreFragment) {
            if (activeFragment == ActiveFragment.EXPLORE) { // Do nothing if same tab
                return true
            }
            exploreFragment = fragment
            activeFragment = ActiveFragment.EXPLORE
        } else if (fragment is BookmarkFragment) {
            if (activeFragment == ActiveFragment.BOOKMARK) { // Do nothing if same tab
                return true
            }
            bookmarkFragment = fragment
            activeFragment = ActiveFragment.BOOKMARK
        } else if (fragment == null && showBottom) {
            if (applicationKvStore!!.getBoolean("login_skipped")
                == true
            ) { // If logged out, more sheet is different
                val bottomSheet = MoreBottomSheetLoggedOutFragment()
                bottomSheet.show(
                    supportFragmentManager,
                    "MoreBottomSheetLoggedOut"
                )
            } else {
                val bottomSheet = MoreBottomSheetFragment()
                bottomSheet.show(
                    supportFragmentManager,
                    "MoreBottomSheet"
                )
            }
        }

        if (fragment != null) {
            supportFragmentManager
                .beginTransaction()
                .replace(R.id.fragmentContainer, fragment)
                .commit()
            return true
        }
        return false
    }

    /**
     * loadFragment() overload that supports passing extras to fragments
     */
    private fun loadFragment(fragment: Fragment?, showBottom: Boolean, args: Bundle?): Boolean {
        if (fragment != null && args != null) {
            fragment.arguments = args
        }

        return loadFragment(fragment, showBottom)
    }

    /**
     * Old implementation of loadFragment() was causing memory leaks, due to MainActivity holding
     * references to cleared fragments. This function frees up all fragment references.
     *
     *
     * Called in loadFragment() before doing the actual loading.
     */
    fun freeUpFragments() {
        // free all fragments except contributionsFragment because several tests depend on it.
        // hence, contributionsFragment is probably still a leak
        nearbyParentFragment = null
        exploreFragment = null
        bookmarkFragment = null
    }


    fun hideTabs() {
        binding!!.fragmentMainNavTabLayout.visibility = View.GONE
    }

    fun showTabs() {
        binding!!.fragmentMainNavTabLayout.visibility = View.VISIBLE
    }

    /**
     * Adds number of uploads next to tab text "Contributions" then it will look like "Contributions
     * (NUMBER)"
     *
     * @param uploadCount
     */
    fun setNumOfUploads(uploadCount: Int) {
        if (activeFragment == ActiveFragment.CONTRIBUTIONS) {
            title =
                resources.getString(R.string.contributions_fragment) + " " + (if (uploadCount != 0)
                    resources
                        .getQuantityString(
                            R.plurals.contributions_subtitle,
                            uploadCount, uploadCount
                        )
                else
                    getString(R.string.contributions_subtitle_zero))
        }
    }

    /**
     * Resume the uploads that got stuck because of the app being killed or the device being
     * rebooted.
     *
     *
     * When the app is terminated or the device is restarted, contributions remain in the
     * 'STATE_IN_PROGRESS' state. This status persists and doesn't change during these events. So,
     * retrieving contributions labeled as 'STATE_IN_PROGRESS' from the database will provide the
     * list of uploads that appear as stuck on opening the app again
     */
    @SuppressLint("CheckResult")
    private fun checkAndResumeStuckUploads() {
        val stuckUploads = contributionDao!!.getContribution(
            listOf(Contribution.STATE_IN_PROGRESS)
        )
            .subscribeOn(Schedulers.io())
            .blockingGet()
        Timber.d("Resuming %d uploads...", stuckUploads.size)
        if (!stuckUploads.isEmpty()) {
            for (contribution in stuckUploads) {
                contribution.state = Contribution.STATE_QUEUED
                contribution.dateUploadStarted = Calendar.getInstance().time
                Completable.fromAction { contributionDao!!.saveSynchronous(contribution) }
                    .subscribeOn(Schedulers.io())
                    .subscribe()
            }
            makeOneTimeWorkRequest(
                this, ExistingWorkPolicy.APPEND_OR_REPLACE
            )
        }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        //quizChecker.initQuizCheck(this);
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("viewPagerCurrentItem", binding!!.pager.currentItem)
        outState.putString("activeFragment", activeFragment!!.name)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        val activeFragmentName = savedInstanceState.getString("activeFragment")
        if (activeFragmentName != null) {
            restoreActiveFragment(activeFragmentName)
        }
    }

    private fun restoreActiveFragment(fragmentName: String) {
        if (fragmentName == ActiveFragment.CONTRIBUTIONS.name) {
            title = getString(R.string.contributions_fragment)
            loadFragment(newInstance(), false)
        } else if (fragmentName == ActiveFragment.NEARBY.name) {
            title = getString(R.string.nearby_fragment)
            loadFragment(NearbyParentFragment.newInstance(), false)
        } else if (fragmentName == ActiveFragment.EXPLORE.name) {
            title = getString(R.string.navigation_item_explore)
            loadFragment(ExploreFragment.newInstance(), false)
        } else if (fragmentName == ActiveFragment.BOOKMARK.name) {
            title = getString(R.string.bookmarks)
            loadFragment(BookmarkFragment.newInstance(), false)
        }
    }

    override fun onBackPressed() {
        when (activeFragment) {
            ActiveFragment.CONTRIBUTIONS -> {
            // Means that contribution fragment is visible
            if (contributionsFragment?.backButtonClicked() != true) { //If this one does not want to handle
                // the back press, let the activity do so
                super.onBackPressed()
                }
            }
        ActiveFragment.NEARBY -> {
            // Means that nearby fragment is visible
            if (nearbyParentFragment?.backButtonClicked() != true) {
            nearbyParentFragment?.let {
                supportFragmentManager.beginTransaction().remove(it).commit()
                    }
                setSelectedItemId(NavTab.CONTRIBUTIONS.code())
                }
            }
         ActiveFragment.EXPLORE -> {
            // Explore Fragment is visible
            if (exploreFragment?.onBackPressed() != true) {
                if (applicationKvStore?.getBoolean("login_skipped") == true) {
                    super.onBackPressed()
                } else {
                    setSelectedItemId(NavTab.CONTRIBUTIONS.code())
                    }
                }
            }
         ActiveFragment.BOOKMARK -> {
            // Means that bookmark fragment is visible
            bookmarkFragment?.onBackPressed()
            }
         else -> {
            super.onBackPressed()
            }
        }
    }

    override fun onBackStackChanged() {
        //initBackButton();
    }

    /**
     * Retry all failed uploads as soon as the user returns to the app
     */
    @SuppressLint("CheckResult")
    private fun retryAllFailedUploads() {
        contributionDao
            ?.getContribution(listOf(Contribution.STATE_FAILED))
            ?.subscribeOn(Schedulers.io())
            ?.subscribe { failedUploads ->
                failedUploads.forEach { contribution ->
                    contributionsFragment?.retryUpload(contribution)
                }
            }
    }

    /**
     * Handles item selection in the options menu. This method is called when a user interacts with
     * the options menu in the Top Bar.
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.upload_tab -> {
                startActivity(Intent(this, UploadProgressActivity::class.java))
                return true
            }

            R.id.notifications -> {
                // Starts notification activity on click to notification icon
                startYourself(this, "unread")
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    fun centerMapToPlace(place: Place?) {
        setSelectedItemId(NavTab.NEARBY.code())
        nearbyParentFragment!!.setNearbyParentFragmentInstanceReadyCallback(
            object : NearbyParentFragmentInstanceReadyCallback {
                override fun onReady() {
                    nearbyParentFragment!!.centerMapToPlace(place)
                }
            })
    }

    /**
     * Launch the Explore fragment from Nearby fragment. This method is called when a user clicks
     * the 'Show in Explore' option in the 3-dots menu in Nearby.
     *
     * @param zoom      current zoom of Nearby map
     * @param latitude  current latitude of Nearby map
     * @param longitude current longitude of Nearby map
     */
    fun loadExploreMapFromNearby(zoom: Double, latitude: Double, longitude: Double) {
        val bundle = Bundle()
        bundle.putDouble("prev_zoom", zoom)
        bundle.putDouble("prev_latitude", latitude)
        bundle.putDouble("prev_longitude", longitude)

        loadFragment(ExploreFragment.newInstance(), false, bundle)
        setSelectedItemId(NavTab.EXPLORE.code())
    }

    /**
     * Launch the Nearby fragment from Explore fragment. This method is called when a user clicks
     * the 'Show in Nearby' option in the 3-dots menu in Explore.
     *
     * @param zoom      current zoom of Explore map
     * @param latitude  current latitude of Explore map
     * @param longitude current longitude of Explore map
     */
    fun loadNearbyMapFromExplore(zoom: Double, latitude: Double, longitude: Double) {
        val bundle = Bundle()
        bundle.putDouble("prev_zoom", zoom)
        bundle.putDouble("prev_latitude", latitude)
        bundle.putDouble("prev_longitude", longitude)

        loadFragment(NearbyParentFragment.newInstance(), false, bundle)
        setSelectedItemId(NavTab.NEARBY.code())
    }

    override fun onResume() {
        super.onResume()

        if ((applicationKvStore!!.getBoolean("firstrun", true)) &&
            (!applicationKvStore!!.getBoolean("login_skipped"))
        ) {
            defaultKvStore.putBoolean("inAppCameraFirstRun", true)
            startWelcome()
        }

        retryAllFailedUploads()
    }

    override fun onDestroy() {
        quizChecker!!.cleanup()
        locationManager!!.unregisterLocationManager()
        // Remove ourself from hashmap to prevent memory leaks
        locationManager = null
        super.onDestroy()
    }

    /**
     * Public method to show nearby from the reference of this.
     */
    fun showNearby() {
        binding!!.fragmentMainNavTabLayout.selectedItemId = NavTab.NEARBY.code()
    }

    enum class ActiveFragment {
        CONTRIBUTIONS,
        NEARBY,
        EXPLORE,
        BOOKMARK,
        MORE
    }

    /**
     * Load default language in onCreate from SharedPreferences
     */
    private fun loadLocale() {
        val preferences = getSharedPreferences(
            "Settings",
            MODE_PRIVATE
        )
        val language = preferences.getString("language", "")!!
        val settingsFragment = SettingsFragment()
        settingsFragment.setLocale(this, language)
    }

    companion object {
        /**
         * Consumers should be simply using this method to use this activity.
         *
         * @param context A Context of the application package implementing this class.
         */
        fun startYourself(context: Context) {
            val intent = Intent(context, MainActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT or Intent.FLAG_ACTIVITY_SINGLE_TOP)
            context.startActivity(intent)
        }
    }
}
